Enable using File Watchers for cloud based Pipeline Roots

Sigmonia opened the pull request:
#### Rationale
https://docs.google.com/document/d/1adCI1KteCV-vNuCJ6Okz1lVcsO6AgZIKXZeSSg2cKQ0/edit?pli=1#

#### Related Pull Requests
* https://github.com/LabKey/cloud/pull/88
* https://github.com/LabKey/commonAssays/pull/373
* https://github.com/LabKey/MaxQuant/pull/40
* https://github.com/LabKey/premium/pull/276
* https://github.com/LabKey/targetedms/pull/435
* https://github.com/LabKey/nlp/pull/199
* https://github.com/LabKey/testAutomation/pull/883

#### Changes
* Lots of File --> nio.Path refactors
* Additional Cloud api interfaces
* new CloudWatcherThread to go along with the FileSystemWatcherImpl

Sigmonia committed:
initial check-in for enabling list reload file watchers to function with S3 and SQS
- Lots of File->Path refactors
- Added Cloud thread to FileWatcher
- Added jCloud SQS and AWS S3 dependencies to cloud Module

TODO: Delete/Clean up SQS Message
 - Throttle CloudWatcherThread

Sigmonia committed:
Fix NPE due to missing/incorrect account name

Sigmonia committed:
Add deleting of originating notification message
- Added call to delete api
- moved delete call to FileSystemWatcherImpl
- Adjusted event stack to pass the messageId through
- reorganize some methods for clarity

Sigmonia committed:
Fixed issue not finding Path key in listenerMap
Fixed issue with deleting temp dir when it isn't empty

Sigmonia committed:
Added timer to polling behavior to manage SQS requests.

Sigmonia committed:
Simplified cloud polling event generation

Sigmonia committed:
Prevent deletion of unsupported/processed messages
Attempt to shutdown timer threads gracefully

Sigmonia committed:
Fix an NPE
Prevent processing error from killing main cloud watcher thread

Sigmonia committed:
Fix Tests failing due to File-->Path conversion

labkey-klum commented on api/src/org/labkey/api/pipeline/PipelineJob.java:
```diff
@@ -433,7 +433,6 @@ public PipeRoot getPipeRoot()
     public void setLogFilePath(Path logFile)
     {
         _logFilePathName = FileUtil.pathToString(logFile);
-//        _logJobStopStart.info("LogFilePath set to " + _logFilePathName);
```
delete?

Sigmonia replied:
deleted.

labkey-klum commented on api/src/org/labkey/api/pipeline/PipelineProvider.java:
```diff
@@ -133,19 +151,20 @@ public boolean accept(File f, boolean checkSiblings)
+                            }
+                            catch (IOException e)
+                            {
                             }
```
should we be throwing an exception or at the least log an error?

Sigmonia replied:
Added Logging and returning false, as I think it better matches the previous behavior.

labkey-klum commented on api/src/org/labkey/api/files/FileSystemWatcherImpl.java:
```diff
@@ -314,6 +319,68 @@ private void close()
+                ExceptionUtil.logExceptionToMothership(null, e);
+            }
+
+            Thread.sleep(10000);
```
Similar to the `FileSystemWatcherThread` would it be possible to move from looping/waiting every 10 seconds? I don't know whether we can get notified by the SQS service and placing some object in a blocking queue.

Sigmonia replied:
This is the difference between SQS and SNS. SQS has no notification system, only polling I did switch it to an explicit timer task. It is now using the same period constant as the pollingQueue.

labkey-klum reviewed (approved):
see inline comments

Sigmonia committed:
Fix error where Import Data button never gets enabled for some file folders.

Sigmonia committed:
CR: add log entry in the event of an error while checking filetypes' filter. Returning false instead of throwing error as I think it better matches previous behavior.

Sigmonia committed:
CR Feedback:
- Switched FileSystemWatcherImpl to use timer for CloudWatchers
- Added better closing/shutdown safeguards
- Cleaned up some lint
- Added logging and comments

Sigmonia committed:
Fix issue with analysis directory not getting set as expected

Sigmonia committed:
Initial fix for timing issue/Race of CloudWatcher thread not starting because module hasn't been loaded yet

Sigmonia committed:
Fix SQS message being deleted when message not processed
Fix an issue where task didn't get cleared and was never processed again.

labkey-tchad committed:
Add actions to mark primary log file similar to error log

Sigmonia committed:
Prevent double URL encoding for log files

Sigmonia committed:
Prevent double URL encoding for log files
updated wrapper CloudPath to include container so it can serialize/deserialize correctly

Sigmonia committed:
Fix listing issue where full file tree was getting listsed, vs just the current level

Sigmonia committed:
Clean up some logging TODOs

Sigmonia committed:
Fix jUnit tests for FileWatcherPipelineTriggerType

Sigmonia committed:
Reset WatchKeys when adding/removing listeners
Update jUnit tests to better look for log files

Sigmonia committed:
Fix test resolution of log file

Sigmonia committed:
Fix a decoding issue for TargetedMS

Sigmonia committed:
Fix unencoded url error for targetedMS

Sigmonia committed:
Remove race condition that may cause NPE

Sigmonia committed:
Adjust cloud schema check.
